content/en/blog/_posts/2021-12-16-StatefulSet-PVC-Auto-Deletion.md

```diff
@@ -0,0 +1,103 @@
+manually or through tools such as Helm, is that the PVCs are tracked by the tool that manages them,
+with explicit lifecycle. Workflows that use StatefulSets must determine on their own what PVCs are
+created by a StatefulSet and what their lifecycle should be.
+
+Before this new feature, when a StatefulSet-managed replica disappears, either because the
+StatefulSet is reducing its replica count, or because its StatefulSet is deleted, the PVC and its
+backing volume remains and must be manually deleted. While this behavior is appropriate when the
+data is critical, in many cases the persistent data in these PVCs is either temporary, or can be
+reconstructed from another source. In those cases, PVCs and their backing volumes remaining after
+their StatefulSet or replicas have been deleted are not necessary, incur cost, and require manual
+cleanup.
+
+## The new StatefulSet PVC retention policy
+
+If you enable the alpha feature, a StatefulSet spec includes a PersistentVolumeClaim retention
+policy.  This is used to control if and when PVCs created from a StatefulSet’s `volumeClaimTemplate`
+are deleted.  This first iteration of the retention policy contains two situations where PVCs may be
+deleted.
+
+The first situation is when the StatefulSet resource is deleted (which implies that all replicas are
+also deleted). This is controlled by the `whenDeleted` policy. The second situation, controlled by
+`whenScaled` is when the StatefulSet is scaled down, which removes some but not all of the replicas
+in a StatefulSet. In both cases the policy can either be `Retain`, where the corresponding PVCs are
+not touched, or `Delete`, which means that PVCs are deleted. The deletion is done with a normal
+[object deletion](/docs/concepts/architecture/garbage-collection/), so that, for example, all
+retention policies for the underlying PV are respected.
+
+This policy forms a matrix with four cases. I’ll walk through and give an example for each one.
+
+  * **`whenDeleted` and `whenScaled` are both `Retain`.** This matches the existing behavior for
+    StatefulSets, where no PVCs are deleted. This is also the default retention policy. It’s
+    appropriate to use when data on StatefulSet volumes may be irreplaceable and should only be
+    deleted manually.
+
+  * **`whenDeleted` is `Delete` and `whenScaled` is `Retain`.** In this case, PVCs are deleted only when
+    the entire StatefulSet is deleted. If the StatefulSet is scaled down, PVCs are not touched,
+    meaning they are available to be reattached if a scale-up occurs with any data from the previous
+    replica. This might be used for a temporary StatefulSet, such as in a CI instance or ETL
+    pipeline, where the data on the StatefulSet is needed only during the lifetime of the
+    StatefulSet lifetime, but while the task is running the data is not easily reconstructible. Any
+    retained state is needed for any replicas that scale down and then up.
```

Comment:
If I Retain an openebs-hostpath PVC on scale-down, how does K8s know to scale-up a new Pod on the same Node as the PVC? My understanding of the creation-time scheduling is that the PVC is created on the Node selected for the Pod, but I'm not clear on how the affinity happens for a PVC that exists before the Pod.

Reply:
I'd have to check that... but since Retain is currently existing behavior, a concise answer is "whatever happens right now" :-)